balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -314,6 +324,64 @@ func deDupAddresses(addrs []resolver.Address) []resolver.Address {
 	return retAddrs
 }
 
+func interleaveAddresses(addrs []resolver.Address) []resolver.Address {
+	// Group the addresses by their type and determine the order in which to
+	// interleave the address families. The order of interleaving the families
+	// is the order in which the first address of a particular type appears in
+	// the address list.
+	familyAddrsMap := map[ipAddrType][]resolver.Address{}
+	interleavingOrder := []ipAddrType{}
+	for _, addr := range addrs {
+		typ := addressType(addr.Addr)
+		if _, found := familyAddrsMap[typ]; !found {
+			interleavingOrder = append(interleavingOrder, typ)
+		}
+		familyAddrsMap[typ] = append(familyAddrsMap[typ], addr)
+	}
+
+	interleavedAddrs := make([]resolver.Address, 0, len(addrs))
+
+	for curTypeIndex := 0; len(interleavedAddrs) < len(addrs); curTypeIndex = (curTypeIndex + 1) % len(interleavingOrder) {
+		// Some IP types may have fewer addresses than others, so we look for
+		// the next type that has a remaining member to add to the interleaved
+		// list.
+		typ := interleavingOrder[curTypeIndex]
+		remainingMembers := familyAddrsMap[typ]
+		if len(remainingMembers) > 0 {
+			interleavedAddrs = append(interleavedAddrs, remainingMembers[0])
+			familyAddrsMap[typ] = remainingMembers[1:]
+		}
+	}
+
+	return interleavedAddrs
+}
+
+func addressType(address string) ipAddrType {
+	// Try parsing addresses without a port specified.
+	ip := net.ParseIP(address)
+	if ip == nil {
+		// Try to parse the IP after removing the port.
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			return ipTypeUnknown
+		}
+		ip = net.ParseIP(host)
+	}
+
+	// If using the passthrough resolver, the hostnames would be unresolved
+	// and therefore not valid IP addresses.
+	if ip == nil {
+		return ipTypeUnknown
+	}
+
+	if ip.To4() != nil {
+		return ipv4
+	} else if ip.To16() != nil {
+		return ipv6
+	}
+	return ipTypeUnknown
```

Comment:
Can we handle this with a `switch` instead?

```golang
switch {
	case ip == nil:
		return ipTypeUnknown
	case ip.To4() != nil:
		return ipv4
	case ip.To16() != nil:
		return ipv6
	default:
		return ipTypeUnknown
}
```

Do we need the `default` case above? i.e. if `ip` is `non-nil` after two attempts at `net.ParseIP`, can we do this instead?
```golang
	if ip == nil {
		return ipTypeUnknown
	}

	if ip.To4() != nil {
		return ipv4
	}
	return ipv6
```